bcda/auth/middleware.go

```diff
@@ -58,48 +59,69 @@ func ParseToken(next http.Handler) http.Handler {
 
 		tokenString := authSubmatches[1]
 
-		token, err := GetProvider().VerifyToken(tokenString)
+		token, ad, err := AuthorizeAccess(tokenString)
 		if err != nil {
-			log.Auth.Errorf("Unable to verify token; %s", err)
 			handleTokenVerificationError(w, rw, err)
 			return
 		}
 
-		var ad AuthData
-		if claims, ok := token.Claims.(*CommonClaims); ok && token.Valid {
-			switch claims.Issuer {
-			case "ssas":
-				ad, err = GetProvider().getAuthDataFromClaims(claims)
-				if err != nil {
-					handleSsasAuthDataError(w, rw, err)
-					return
-				}
-			default:
-				log.Auth.Errorf("Unsupported claims issuer %s", claims.Issuer)
-				rw.Exception(w, http.StatusNotFound, responseutils.TokenErr, "")
-				return
-			}
-		}
 		ctx := context.WithValue(r.Context(), TokenContextKey, token)
 		ctx = context.WithValue(ctx, AuthDataContextKey, ad)
 		next.ServeHTTP(w, r.WithContext(ctx))
 	})
 }
 
-func handleSsasAuthDataError(w http.ResponseWriter, rw fhirResponseWriter, err error) {
-	log.Auth.Error(err)
-	if _, ok := err.(*customErrors.EntityNotFoundError); ok {
-		rw.Exception(w, http.StatusForbidden, responseutils.UnauthorizedErr, responseutils.UnknownEntityErr)
-	} else {
-		rw.Exception(w, http.StatusUnauthorized, responseutils.TokenErr, "")
+// AuthorizeAccess asserts that a base64 encoded token string is valid for accessing the BCDA API.
+func AuthorizeAccess(tokenString string) (*jwt.Token, AuthData, error) {
+	tknEvent := event{op: "AuthorizeAccess"}
+	operationStarted(tknEvent)
+	token, err := GetProvider().VerifyToken(tokenString)
+
+	var ad AuthData
+
+	if err != nil {
+		tknEvent.help = fmt.Sprintf("VerifyToken failed in AuthorizeAccess; %s", err.Error())
+		operationFailed(tknEvent)
+		return nil, ad, err
+	}
+
+	claims, ok := token.Claims.(*CommonClaims)
+	if !ok || !token.Valid {
```

Comment:
The `token.Valid` check is new, pulled from `parseToken`. Note however that VerifyToken always sets token.Valid to true, so this check should always pass.

Reply:
Note also that this introduces one small functional change to unauthenticated endpoints. Previously if this check failed, `ParseToken` would not fail the unauthenticated request. Now the request will fail.

I checked in the code to confirm that this is expected to be unreachable code, because...

- [VerifyToken already checks for token.Claims](https://github.com/CMSgov/bcda-app/blob/dd356ecda6aa85fb9befddc02b09a9e8f7e7907c/bcda/auth/ssas.go#L238-L242)
- [VerifyToken already sets token.Valid to true](https://github.com/CMSgov/bcda-app/blob/dd356ecda6aa85fb9befddc02b09a9e8f7e7907c/bcda/auth/ssas.go#L208)

Reply:
What would happen previously from the consumer side? I see in the old code that it has a `return error` statement.